datafusion/physical-expr/src/expressions/dynamic.rs

```diff
@@ -0,0 +1,115 @@
+
+use datafusion_expr::ColumnarValue;
+use datafusion_physical_expr_common::physical_expr::PhysicalExpr;
+use datafusion_common::error::Result;
+
+/// A source for dynamically generated PhysicalExprs.
+/// Operators may want to dynamically generate a PhysicalExpr lazily at runtime.
+/// For example, a TopK operator can push down it's current state into a join operator
+/// to dynamically filter rows that would not be in the TopK anyway as the join takes place.
+pub trait DynamicPhysicalExprSource: Send + Sync + std::fmt::Debug + 'static {
+    /// Returns a PhysicalExpr that may be dynamically generated by the implementer.
+    fn current_expr(&self) -> Result<Arc<dyn PhysicalExpr>>;
+}
+
+#[derive(Debug)]
+pub struct DynamicPhysicalExpr {
+    current: Arc<dyn PhysicalExpr>,
+    inner: Arc<dyn DynamicPhysicalExprSource>,
+}
+
+impl std::fmt::Display for DynamicPhysicalExpr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Ok(expr) = self.inner.current_expr() {
+            write!(f, "DynamicPhysicalExpr({})", expr)
+        } else {
+            write!(f, "DynamicPhysicalExpr(?)")
+        }
+    }
+}
+
+impl DynamicPhysicalExpr {
+    pub fn new(inner: Arc<dyn DynamicPhysicalExprSource>) -> Result<Self> {
+        let current = inner.current_expr()?;
+        Ok(Self { current, inner })
+    }
+
+    pub fn current(&self) -> Arc<dyn PhysicalExpr> {
+        self.inner.current_expr().unwrap_or_else(|_| self.current.clone())
+    }
+}
+
+
+// Manually derive PartialEq and Hash to work around https://github.com/rust-lang/rust/issues/78808
+impl PartialEq for DynamicPhysicalExpr {
+    fn eq(&self, other: &Self) -> bool {
+        self.current().eq(&other.current())
+    }
+}
+
+impl Eq for DynamicPhysicalExpr {}
+
+impl Hash for DynamicPhysicalExpr {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.current().hash(state)
+    }
+}
+
+
+impl PhysicalExpr for DynamicPhysicalExpr {
+    /// Return a reference to Any that can be used for downcasting
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn children(&self) -> Vec<&Arc<dyn PhysicalExpr>> {
+        // Can't get children, returning an &Arc would require returning a reference to a temporary value
+        vec![]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn PhysicalExpr>>,
+    ) -> Result<Arc<dyn PhysicalExpr>> {
+        // Do we need to do something here?
+        Ok(self)
+    }
+
+    fn data_type(&self, input_schema: &arrow::datatypes::Schema) -> Result<arrow::datatypes::DataType> {
+        self.inner.current_expr()?.data_type(input_schema)
+    }
+
+    fn nullable(&self, input_schema: &arrow::datatypes::Schema) -> Result<bool> {
+        self.inner.current_expr()?.nullable(input_schema)
+    }
+
+    fn evaluate(&self, batch: &arrow::record_batch::RecordBatch) -> Result<ColumnarValue> {
+        self.inner.current_expr()?.evaluate(batch)
```

Comment:
This implementation updates the bounds on each call to evaluate - we could potentially optimize it so it updated every 10 or 100 calls for example, to amortize the cost of the updated bounds check

Reply:
Perhaps this i what you meant by https://github.com/apache/datafusion/pull/15301#issuecomment-2760193801

> Possible performance / lock contention -> this is probably solvable by balancing update frequency.

Reply:
Yes exactly. But I feel like arbitrarily doing it every X calls is probably not ideal. I'd guess doing it every "unit of compute" makes more sense. E.g. update it for each batch.

Reply:
Which kind of gets us back to having a "first class" `DynamicFiltersSource` that's not hidden behind the PhysicalExpr APIs?

Like I feel that the "bottleneck" for updating it with the current design is going to be TopK itself getting new data to be able to update. Be it a `PhysicalExpr` or using `DynamicFilterSource` directly we're still going to want to manually decide when to refresh it (before opening each file in DataSourceExec, before processing each batch in a join/filter exec, etc.).

Reply:
> Yes exactly. But I feel like arbitrarily doing it every X calls is probably not ideal. I'd guess doing it every "unit of compute" makes more sense. E.g. update it for each batch.

Maybe the topk could remember / track when its value changes and we could trigger updating the predicate whenever new information was found